### PR TITLE
Fix logging and end of training logic, and add fine-tuning instructions

### PR DESCRIPTION
**Pull request by heewooj**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -98,17 +98,32 @@ mpiexec -n {ngpus} python jukebox/train.py --hps=small_vqvae,small_upsampler,all
...
+mpiexec -n {ngpus} python jukebox/train.py --hps=vqvae,small_prior,all_fp16,cpu_ema --name=pretrained_vqvae_small_prior --sample_length=1048576 --bs=4 --nworkers=4 --bs_sample=4 --aug_shift --aug_blend --audio_files_dir={audio_files_dir} --labels=False --train --test --prior --levels=3 --level=2 --weight_decay=0.01 --save_iters=1000 --lr_use_linear_decay --lr_decay={decay_steps_as_needed}
 ```
 
 You can then run sample.py with the top-level of our models replaced by your new model. To do so, add an entry `my_model` in MODELs (in `make_models.py`) with the (vqvae hps, upsampler hps, top-level prior hps) of your new model, and run sample.py with `--model=my_model`. 
-	
+
+Training the `small_prior` with a batch size of 2, 4, and 8 requires 6.7 GB, 9.3 GB, and 15.8 GB of GPU memory, respectively. A few days to a week of training typically yields reasonable samples when the dataset is homogeneous (e.g. all piano pieces, songs of the same style, etc).
+
+### Fine-tune pre-trained top-level prior to new style(s)
+
```
Comment (prafullasd) [inline README.md:115]: New README has a section on how to deal with labels/metadata, so maybe we can put this after that section? And update this readme accordingly?

File: jukebox/train.py [hunk 2]
```diff
@@ -31,9 +33,13 @@ def log_aud(logger, tag, x, hps):
...
     y = y.cpu().numpy()
-    txt = f''
+    attrs = ["artist", "genre", "lyrics"]
+    labels = []
     for item in range(y.shape[0]):
-        txt += labeller.describe_label(y)
+        meta = labeller.describe_label(y[item])
+        meta = [("sample", item), *list(map(lambda attr: (attr, meta[attr]), attrs))]
+        labels.append(json.dumps(OrderedDict(meta)))
+    txt = "\n".join(labels)
```
Comment (prafullasd) [inline jukebox/train.py:42]: fix merged in previous pull request, so you might need to resolve merge conflict here

File: README.md (hunk 1 above)
Comment (prafullasd) [reply README.md]: I just did a merge conflict resolve, see if it looks ok?

File: README.md [hunk 3]
```diff
@@ -194,6 +196,21 @@ This uses `attn_order=12` which includes `prime_attention` layers with keys/valu
...
+1. Support `--labels=True` by implementing `get_metadata` in `jukebox/data/files_dataset.py` for your dataset.
+2. Add new entries in `jukebox/data/ids`. 
+  * We recommend replacing existing mappings (e.g. rename `"unknown"`, etc with styles of your choice; you will have to create a new branch for this). This uses the pre-trained style vectors as initialization and could potentially save some compute.
+3. Run
+```
+mpiexec -n {ngpus} python jukebox/train.py --hps=vqvae,prior_1b_lyrics,all_fp16,cpu_ema --name=finetuned \
+--sample_length=1048576 --bs=1 --aug_shift --aug_blend --audio_files_dir={audio_files_dir} \
+--labels=True --train --test --prior --levels=3 --level=2 --weight_decay=0.01 --save_iters=1000 \
+--min_duration=17.84 --max_duration=600 --c_res=1 --lr_use_linear_decay --lr_decay={decay_steps_as_needed}
+```
```
Comment (prafullasd) [inline README.md:211]: I followed the convention to change these hps directly in the dictionary in hparams.py so that they are correctly used when we call make_models.py for sampling. You might want to do same here?

Comment (heewooj) [reply README.md:225]: sg! some default hps values are moved to the dictionary now

File: README.md [hunk 4]
```diff
@@ -125,18 +125,20 @@ checkpoint and run with
...
 
-To retrain top-level on a new dataset, run
+To train top-level on a new dataset, run
 ```
 mpiexec -n {ngpus} python jukebox/train.py --hps=vqvae,small_prior,all_fp16,cpu_ema --name=pretrained_vqvae_small_prior \
 --sample_length=1048576 --bs=4 --aug_shift --aug_blend --audio_files_dir={audio_files_dir} \
---labels=False --train --test --prior --levels=3 --level=2 --weight_decay=0.01 --save_iters=1000
+--labels=False --train --test --prior --levels=3 --level=2 --weight_decay=0.01 --save_iters=1000 \
+--lr_use_linear_decay --lr_decay={decay_steps_as_needed}
 ```
```
Comment (prafullasd) [inline README.md:140]: Lets remove linear decay here, since I added a separate step for it in the comment in line 143

File: jukebox/train.py [hunk 5]
```diff
@@ -31,11 +31,11 @@ def log_aud(logger, tag, x, hps):
 
 def log_labels(logger, labeller, tag, y, hps):
     y = y.cpu().numpy()
-    txt = f''
+    txt = ''
     for item in range(y.shape[0]):
-        description = labeller.describe_label(y)
+        description = labeller.describe_label(y[item])
```
Comment (prafullasd) [inline jukebox/train.py:36]: good catch!

File: README.md (hunk 1 above)
Comment (heewooj) [reply README.md:212]: looks good

File: jukebox/hparams.py [hunk 6]
```diff
@@ -140,7 +140,8 @@ def setup_hparams(hparam_set_names, kwargs):
     prime_blocks=32,
     prime_init_scale=0.7,
     prime_c_res=1,
-    t_ranges=((1049580.0, 26460000.0), (0.0, 26460000.0), (0.0, 1.0)),
+    min_duration=23.8,
+    max_duration=600.0,
     use_tokens=True,
     n_tokens=512,
     prime_loss_fraction=0.4,
```
Comment (prafullasd) [inline jukebox/hparams.py:147]: Also should put the values for 5b mode

File: README.md (hunk 4 above)
Comment (heewooj) [reply README.md:140]: ah okay

File: README.md (hunk 3 above)
Comment (prafullasd) [reply README.md:225]: Same for linear decay here as above?

File: jukebox/hparams.py (hunk 6 above)
Comment (heewooj) [reply jukebox/hparams.py:148]: good catch

File: jukebox/hparams.py [hunk 7]
```diff
@@ -170,8 +172,10 @@ def setup_hparams(hparam_set_names, kwargs):
     attn_order=12,
     blocks=64,
     init_scale=0.2,
+    c_res=1,
```
Comment (prafullasd) [inline jukebox/hparams.py:175]: Would this affect sampling?

Comment (heewooj) [reply jukebox/hparams.py:175]: good q, checkpointing is set for training
